Quarkus Misc2 tests fail with 20.3

## Description
`quarkus-integration-test-tika` fails to build with:
```
2020-11-17T20:19:36.3437448Z Error: Unsupported features in 3 methods
2020-11-17T20:19:36.3442540Z Detailed message:
2020-11-17T20:19:36.3448425Z Error: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
2020-11-17T20:19:36.3449967Z Trace: Object was reached by 
2020-11-17T20:19:36.3450888Z 	reading field java.awt.image.ColorModel.colorSpace of
2020-11-17T20:19:36.3452024Z 		constant java.awt.image.DirectColorModel@5984b36 reached by 
2020-11-17T20:19:36.3453161Z 	reading field java.awt.image.ColorModel.RGBdefault
2020-11-17T20:19:36.3455353Z Error: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
2020-11-17T20:19:36.3456878Z Trace: Object was reached by 
2020-11-17T20:19:36.3458514Z 	reading field org.apache.pdfbox.pdmodel.graphics.color.PDDeviceRGB.awtColorSpace of
2020-11-17T20:19:36.3460897Z 		constant org.apache.pdfbox.pdmodel.graphics.color.PDDeviceRGB@5eecfc90 reached by 
2020-11-17T20:19:36.3464161Z 	scanning method org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(PDICCBased.java:290)
2020-11-17T20:19:36.3468235Z Call path from entry point to org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(Exception): 
2020-11-17T20:19:36.3472907Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(PDICCBased.java:288)
2020-11-17T20:19:36.3476417Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.loadICCProfile(PDICCBased.java:266)
2020-11-17T20:19:36.3479107Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.<init>(PDICCBased.java:131)
2020-11-17T20:19:36.3481412Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.create(PDICCBased.java:160)
2020-11-17T20:19:36.3483928Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:192)
2020-11-17T20:19:36.3490008Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:70)
2020-11-17T20:19:36.3492574Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:53)
2020-11-17T20:19:36.3495319Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDDeviceNProcess.getColorSpace(PDDeviceNProcess.java:75)
2020-11-17T20:19:36.3498271Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDDeviceNProcess.toString(PDDeviceNProcess.java:103)
2020-11-17T20:19:36.3500182Z 	at java.base@11.0.9/java.lang.String.valueOf(String.java:2951)
2020-11-17T20:19:36.3501323Z 	at java.base@11.0.9/java.io.PrintStream.print(PrintStream.java:745)
2020-11-17T20:19:36.3502441Z 	at java.base@11.0.9/java.io.PrintStream.println(PrintStream.java:882)
2020-11-17T20:19:36.3504206Z 	at app//com.oracle.svm.jni.functions.JNIFunctions.ExceptionDescribe(JNIFunctions.java:757)
2020-11-17T20:19:36.3506834Z 	at com.oracle.svm.core.code.IsolateEnterStub.JNIFunctions_ExceptionDescribe_b5412f7570bccae90b000bc37855f00408b2ad73(generated:0)
2020-11-17T20:19:36.3511294Z Error: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
2020-11-17T20:19:36.3513848Z Trace: 
2020-11-17T20:19:36.3516435Z 	at parsing org.apache.pdfbox.pdmodel.graphics.color.PDCIEDictionaryBasedColorSpace.convXYZtoRGB(PDCIEDictionaryBasedColorSpace.java:91)
2020-11-17T20:19:36.3521093Z Call path from entry point to org.apache.pdfbox.pdmodel.graphics.color.PDCIEDictionaryBasedColorSpace.convXYZtoRGB(float, float, float): 
2020-11-17T20:19:36.3525798Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDCIEDictionaryBasedColorSpace.convXYZtoRGB(PDCIEDictionaryBasedColorSpace.java:79)
2020-11-17T20:19:36.3529132Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDLab.toRGB(PDLab.java:124)
2020-11-17T20:19:36.3531095Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColor.toRGB(PDColor.java:153)
2020-11-17T20:19:36.3533357Z 	at org.apache.pdfbox.rendering.TilingPaintFactory$TilingPaintParameter.equals(TilingPaintFactory.java:131)
2020-11-17T20:19:36.3535475Z 	at java.base@11.0.9/java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:940)
2020-11-17T20:19:36.3537032Z 	at java.base@11.0.9/java.util.Properties.getProperty(Properties.java:1125)
2020-11-17T20:19:36.3539072Z 	at app//com.oracle.svm.core.jdk.SystemPropertiesSupport.getProperty(SystemPropertiesSupport.java:144)
2020-11-17T20:19:36.3541345Z 	at app//com.oracle.svm.core.jdk.Target_java_lang_System.getProperty(JavaLangSubstitutions.java:330)
2020-11-17T20:19:36.3543325Z 	at com.oracle.svm.jni.JNIJavaCallWrappers.jniInvoke_VA_LIST:Ljava_lang_System_2_0002egetProperty_00028Ljava_lang_String_2_00029Ljava_lang_String_2(generated:0)
2020-11-17T20:19:36.3544532Z 
2020-11-17T20:19:36.3545424Z com.oracle.svm.core.util.UserError$UserException: Unsupported features in 3 methods
2020-11-17T20:19:36.3546361Z Detailed message:
2020-11-17T20:19:36.3548243Z Error: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
2020-11-17T20:19:36.3550000Z Trace: Object was reached by 
2020-11-17T20:19:36.3550877Z 	reading field java.awt.image.ColorModel.colorSpace of
2020-11-17T20:19:36.3552010Z 		constant java.awt.image.DirectColorModel@5984b36 reached by 
2020-11-17T20:19:36.3553156Z 	reading field java.awt.image.ColorModel.RGBdefault
2020-11-17T20:19:36.3555445Z Error: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
2020-11-17T20:19:36.3556957Z Trace: Object was reached by 
2020-11-17T20:19:36.3558595Z 	reading field org.apache.pdfbox.pdmodel.graphics.color.PDDeviceRGB.awtColorSpace of
2020-11-17T20:19:36.3561013Z 		constant org.apache.pdfbox.pdmodel.graphics.color.PDDeviceRGB@5eecfc90 reached by 
2020-11-17T20:19:36.3564056Z 	scanning method org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(PDICCBased.java:290)
2020-11-17T20:19:36.3568431Z Call path from entry point to org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(Exception): 
2020-11-17T20:19:36.3572398Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(PDICCBased.java:288)
2020-11-17T20:19:36.3575815Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.loadICCProfile(PDICCBased.java:266)
2020-11-17T20:19:36.3578249Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.<init>(PDICCBased.java:131)
2020-11-17T20:19:36.3580466Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.create(PDICCBased.java:160)
2020-11-17T20:19:36.3582854Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:192)
2020-11-17T20:19:36.3585168Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:70)
2020-11-17T20:19:36.3587494Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:53)
2020-11-17T20:19:36.3590202Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDDeviceNProcess.getColorSpace(PDDeviceNProcess.java:75)
2020-11-17T20:19:36.3593137Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDDeviceNProcess.toString(PDDeviceNProcess.java:103)
2020-11-17T20:19:36.3594949Z 	at java.base@11.0.9/java.lang.String.valueOf(String.java:2951)
2020-11-17T20:19:36.3596154Z 	at java.base@11.0.9/java.io.PrintStream.print(PrintStream.java:745)
2020-11-17T20:19:36.3597242Z 	at java.base@11.0.9/java.io.PrintStream.println(PrintStream.java:882)
2020-11-17T20:19:36.3599544Z 	at app//com.oracle.svm.jni.functions.JNIFunctions.ExceptionDescribe(JNIFunctions.java:757)
2020-11-17T20:19:36.3602125Z 	at com.oracle.svm.core.code.IsolateEnterStub.JNIFunctions_ExceptionDescribe_b5412f7570bccae90b000bc37855f00408b2ad73(generated:0)
2020-11-17T20:19:36.3646258Z Error: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
2020-11-17T20:19:36.3648950Z Trace: 
2020-11-17T20:19:36.3651618Z 	at parsing org.apache.pdfbox.pdmodel.graphics.color.PDCIEDictionaryBasedColorSpace.convXYZtoRGB(PDCIEDictionaryBasedColorSpace.java:91)
2020-11-17T20:19:36.3655901Z Call path from entry point to org.apache.pdfbox.pdmodel.graphics.color.PDCIEDictionaryBasedColorSpace.convXYZtoRGB(float, float, float): 
2020-11-17T20:19:36.3660512Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDCIEDictionaryBasedColorSpace.convXYZtoRGB(PDCIEDictionaryBasedColorSpace.java:79)
2020-11-17T20:19:36.3663872Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDLab.toRGB(PDLab.java:124)
2020-11-17T20:19:36.3665839Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColor.toRGB(PDColor.java:153)
2020-11-17T20:19:36.3668075Z 	at org.apache.pdfbox.rendering.TilingPaintFactory$TilingPaintParameter.equals(TilingPaintFactory.java:131)
2020-11-17T20:19:36.3670223Z 	at java.base@11.0.9/java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:940)
2020-11-17T20:19:36.3671987Z 	at java.base@11.0.9/java.util.Properties.getProperty(Properties.java:1125)
2020-11-17T20:19:36.3674051Z 	at app//com.oracle.svm.core.jdk.SystemPropertiesSupport.getProperty(SystemPropertiesSupport.java:144)
2020-11-17T20:19:36.3676388Z 	at app//com.oracle.svm.core.jdk.Target_java_lang_System.getProperty(JavaLangSubstitutions.java:330)
2020-11-17T20:19:36.3678413Z 	at com.oracle.svm.jni.JNIJavaCallWrappers.jniInvoke_VA_LIST:Ljava_lang_System_2_0002egetProperty_00028Ljava_lang_String_2_00029Ljava_lang_String_2(generated:0)
2020-11-17T20:19:36.3679533Z 
2020-11-17T20:19:36.3680314Z 	at com.oracle.svm.core.util.UserError.abort(UserError.java:82)
2020-11-17T20:19:36.3681978Z 	at com.oracle.svm.hosted.FallbackFeature.reportAsFallback(FallbackFeature.java:217)
2020-11-17T20:19:36.3684342Z 	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:768)
2020-11-17T20:19:36.3686714Z 	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:558)
2020-11-17T20:19:36.3688699Z 	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:471)
2020-11-17T20:19:36.3690517Z 	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
2020-11-17T20:19:36.3692126Z 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
2020-11-17T20:19:36.3693819Z 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
2020-11-17T20:19:36.3695423Z 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
2020-11-17T20:19:36.3697006Z 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
2020-11-17T20:19:36.3698889Z 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
2020-11-17T20:19:36.3701362Z Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Unsupported features in 3 methods
2020-11-17T20:19:36.3703065Z Detailed message:
2020-11-17T20:19:36.3705133Z Error: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
2020-11-17T20:19:36.3706649Z Trace: Object was reached by 
2020-11-17T20:19:36.3707502Z 	reading field java.awt.image.ColorModel.colorSpace of
2020-11-17T20:19:36.3708661Z 		constant java.awt.image.DirectColorModel@5984b36 reached by 
2020-11-17T20:19:36.3709768Z 	reading field java.awt.image.ColorModel.RGBdefault
2020-11-17T20:19:36.3711964Z Error: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
2020-11-17T20:19:36.3713446Z Trace: Object was reached by 
2020-11-17T20:19:36.3715116Z 	reading field org.apache.pdfbox.pdmodel.graphics.color.PDDeviceRGB.awtColorSpace of
2020-11-17T20:19:36.3717488Z 		constant org.apache.pdfbox.pdmodel.graphics.color.PDDeviceRGB@5eecfc90 reached by 
2020-11-17T20:19:36.3721960Z 	scanning method org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(PDICCBased.java:290)
2020-11-17T20:19:36.3726225Z Call path from entry point to org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(Exception): 
2020-11-17T20:19:36.3730242Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.fallbackToAlternateColorSpace(PDICCBased.java:288)
2020-11-17T20:19:36.3733616Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.loadICCProfile(PDICCBased.java:266)
2020-11-17T20:19:36.3736091Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.<init>(PDICCBased.java:131)
2020-11-17T20:19:36.3738269Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDICCBased.create(PDICCBased.java:160)
2020-11-17T20:19:36.3740707Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:192)
2020-11-17T20:19:36.3743470Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:70)
2020-11-17T20:19:36.3745968Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColorSpace.create(PDColorSpace.java:53)
2020-11-17T20:19:36.3749037Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDDeviceNProcess.getColorSpace(PDDeviceNProcess.java:75)
2020-11-17T20:19:36.3752200Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDDeviceNProcess.toString(PDDeviceNProcess.java:103)
2020-11-17T20:19:36.3754181Z 	at java.base@11.0.9/java.lang.String.valueOf(String.java:2951)
2020-11-17T20:19:36.3755170Z 	at java.base@11.0.9/java.io.PrintStream.print(PrintStream.java:745)
2020-11-17T20:19:36.3756285Z 	at java.base@11.0.9/java.io.PrintStream.println(PrintStream.java:882)
2020-11-17T20:19:36.3758049Z 	at app//com.oracle.svm.jni.functions.JNIFunctions.ExceptionDescribe(JNIFunctions.java:757)
2020-11-17T20:19:36.3760616Z 	at com.oracle.svm.core.code.IsolateEnterStub.JNIFunctions_ExceptionDescribe_b5412f7570bccae90b000bc37855f00408b2ad73(generated:0)
2020-11-17T20:19:36.3764866Z Error: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: No instances of java.awt.color.ICC_ColorSpace are allowed in the image heap as this class should be initialized at image runtime. To see how this object got instantiated use --trace-object-instantiation=java.awt.color.ICC_ColorSpace.
2020-11-17T20:19:36.3767513Z Trace: 
2020-11-17T20:19:36.3770130Z 	at parsing org.apache.pdfbox.pdmodel.graphics.color.PDCIEDictionaryBasedColorSpace.convXYZtoRGB(PDCIEDictionaryBasedColorSpace.java:91)
2020-11-17T20:19:36.3774749Z Call path from entry point to org.apache.pdfbox.pdmodel.graphics.color.PDCIEDictionaryBasedColorSpace.convXYZtoRGB(float, float, float): 
2020-11-17T20:19:36.3779282Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDCIEDictionaryBasedColorSpace.convXYZtoRGB(PDCIEDictionaryBasedColorSpace.java:79)
2020-11-17T20:19:36.3782661Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDLab.toRGB(PDLab.java:124)
2020-11-17T20:19:36.3784593Z 	at org.apache.pdfbox.pdmodel.graphics.color.PDColor.toRGB(PDColor.java:153)
2020-11-17T20:19:36.3786861Z 	at org.apache.pdfbox.rendering.TilingPaintFactory$TilingPaintParameter.equals(TilingPaintFactory.java:131)
2020-11-17T20:19:36.3788990Z 	at java.base@11.0.9/java.util.concurrent.ConcurrentHashMap.get(ConcurrentHashMap.java:940)
2020-11-17T20:19:36.3790585Z 	at java.base@11.0.9/java.util.Properties.getProperty(Properties.java:1125)
2020-11-17T20:19:36.3792611Z 	at app//com.oracle.svm.core.jdk.SystemPropertiesSupport.getProperty(SystemPropertiesSupport.java:144)
2020-11-17T20:19:36.3794957Z 	at app//com.oracle.svm.core.jdk.Target_java_lang_System.getProperty(JavaLangSubstitutions.java:330)
2020-11-17T20:19:36.3796812Z 	at com.oracle.svm.jni.JNIJavaCallWrappers.jniInvoke_VA_LIST:Ljava_lang_System_2_0002egetProperty_00028Ljava_lang_String_2_00029Ljava_lang_String_2(generated:0)
2020-11-17T20:19:36.3797925Z 
2020-11-17T20:19:36.3799426Z 	at com.oracle.graal.pointsto.constraints.UnsupportedFeatures.report(UnsupportedFeatures.java:129)
2020-11-17T20:19:36.3802087Z 	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:765)
2020-11-17T20:19:36.3803595Z 	... 8 more
2020-11-17T20:19:36.3957861Z Error: Image build request failed with exit status 1
```

## How To Reproduce
Steps to reproduce the behavior:
1. Build `mandrel/20.3` using https://github.com/graalvm/mandrel-packaging
2. `git clone git@github.com:quarkusio/quarkus.git`
3. `cd quarkus`
3. `mvn -Dnative.surefire.skip -Dnative verify -pl integration-tests/tika`

### Configuration:
 - OS: Fedora 33
 - Architecture: AMD64
 - Mandrel version: `mandrel/20.3` branch
 - JDK version: OpenJDK 11.0.9+11

## Expected behavior
Tests should pass

## Additional context
At build time the following warnings are printed:
```
2020-11-17T20:15:37.4008400Z 20:15:37,390 WARNING [org.apa.tik.par.PDFParser] J2KImageReader not loaded. JPEG2000 files will not be processed.
2020-11-17T20:15:37.4011163Z See https://pdfbox.apache.org/2.0/dependencies.html#jai-image-io
2020-11-17T20:15:37.4012413Z for optional dependencies.
```

--- a/substratevm/mx.substratevm/mx_substratevm.py
+++ b/substratevm/mx.substratevm/mx_substratevm.py
@@ -1420,8 +1420,6 @@ class SubstrateCompilerFlagsBuilder(mx.ArchivableProject):
                 # Reflective access to sun.security.x509.OIDMap.nameMap
                 'java.base/sun.security.x509',
                 'java.base/jdk.internal.logger',
-                # Resource registration of resources under sun.java2d.cmm.profiles
-                'java.desktop/sun.java2d.cmm.profiles',
 
                 # Reflective access to org.graalvm.nativeimage.impl.ImageSingletonsSupport.
                 'org.graalvm.sdk/org.graalvm.nativeimage.impl',

--- a/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaAWTSubstitutions.java
+++ b/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaAWTSubstitutions.java
@@ -35,7 +35,7 @@ import java.awt.GraphicsEnvironment;
 import java.io.FilenameFilter;
 import java.util.function.BooleanSupplier;
 
-@Platforms(Platform.LINUX_AMD64.class)
+@Platforms(Platform.LINUX.class)
 @SuppressWarnings({"static-method", "unused"})
 public final class JavaAWTSubstitutions {
     // Checkstyle: stop

--- a/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaAWTSubstitutions.java
+++ b/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/JavaAWTSubstitutions.java
@@ -24,36 +24,516 @@
  */
 package com.oracle.svm.core.jdk;
 
-import java.awt.Toolkit;
-
-import com.oracle.svm.core.annotate.Alias;
-import com.oracle.svm.core.annotate.RecomputeFieldValue;
+import com.oracle.svm.core.annotate.Delete;
 import com.oracle.svm.core.annotate.Substitute;
 import com.oracle.svm.core.annotate.TargetClass;
+import org.graalvm.compiler.serviceprovider.JavaVersionUtil;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
 
-@TargetClass(java.awt.Toolkit.class)
-final class Target_java_awt_Toolkit {
+import java.awt.GraphicsEnvironment;
+import java.io.FilenameFilter;
+import java.util.function.BooleanSupplier;
 
-    @Substitute
-    private static Toolkit getDefaultToolkit() {
-        throw new IllegalArgumentException("AWT is currently not supported on Substrate VM");
-    }
-}
-
-/** Dummy class to have a class with the file's name. */
+@Platforms(Platform.LINUX_AMD64.class)
+@SuppressWarnings({"static-method", "unused"})
 public final class JavaAWTSubstitutions {
-}
+    // Checkstyle: stop
+    @TargetClass(className = "sun.awt.DebugSettings", onlyWith = JDK11OrLater.class)
+    static final class Target_sun_awt_DebugSettings {
 
-/**
- * The purpose of this substitution is to prevent standard color profile loading deferral. The only
- * reason this substitution exists is to ensure that our fix for loading standard color profiles
- * works. It should be removed once standard color profile loading is tested.
- */
-@TargetClass(className = "sun.java2d.cmm.ProfileDeferralMgr", onlyWith = JDK11OrLater.class)
-final class Target_sun_java2d_cmm_ProfileDeferralMgr {
+        // Could not link with Java_sun_awt_DebugSettings_setCTracingOn__Z JNI method
+        @Substitute
+        void setCTracingOn(boolean enabled) {
+            throw new UnsupportedOperationException();
+        }
 
-    @Alias//
-    @RecomputeFieldValue(kind = RecomputeFieldValue.Kind.FromAlias)//
-    public static boolean deferring = false;
+        // Could not link with
+        // Java_sun_awt_DebugSettings_setCTracingOn_JDK11OrLater.class_ZLjava_lang_String_2 JNI
+        // method
+        @Substitute
+        void setCTracingOn(boolean enabled, String file) {
+            throw new UnsupportedOperationException();
+        }
 
+        // Could not link with Java_sun_awt_DebugSettings_setCTracingOn__ZLjava_lang_String_2I JNI
+        // method
+        @Substitute
+        void setCTracingOn(boolean enabled, String file, int line) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.java2d.loops.TransformBlit", onlyWith = JDK11OrLater.class)
+    static final class Target_sun_java2d_loops_TransformBlit {
+
+        // Could not find JNI method Java_sun_java2d_loops_TransformBlit_Transform
+        @Substitute
+        void Transform(sun.java2d.SurfaceData src, sun.java2d.SurfaceData dst,
+                        java.awt.Composite comp, sun.java2d.pipe.Region clip,
+                        java.awt.geom.AffineTransform at, int hint,
+                        int srcx, int srcy, int dstx, int dsty,
+                        int width, int height) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.font.FileFontStrike", onlyWith = JDK11OrLater.class)
+    static final class Target_sun_font_FileFontStrike {
+
+        // Java_sun_font_FileFontStrike_initNative belongs to Windows static lib
+        @Substitute
+        static boolean initNative() {
+            throw new UnsupportedOperationException();
+        }
+
+        // Java_sun_font_FileFontStrike_initNative belongs to Windows static lib
+        @Substitute
+        long _getGlyphImageFromWindows(String family,
+                        int style,
+                        int size,
+                        int glyphCode,
+                        boolean fracMetrics,
+                        int fontDataSize) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.awt.FontConfiguration", onlyWith = JDK11OrLater.class)
+    static final class Target_sun_awt_FontConfiguration {
+
+        // To prevent an attempt to load fonts from java.home
+        @Substitute
+        public boolean foundOsSpecificFile() {
+            return false;
+        }
+    }
+
+    /*
+     * To prevent AWT linkage error that happens with 'awt_headless' in headless mode, we substitute
+     * native methods that depend on 'awt_xawt' library in the call-tree.
+     */
+    @TargetClass(className = "sun.awt.X11.XToolkit", onlyWith = IsHeadless.class)
+    static final class Target_sun_awt_X11_XToolkit {
+
+        @Substitute
+        static String getEnv(String key) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static void wakeup_poll() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        void run() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        int getNumberOfButtons() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "java.awt.Window", onlyWith = IsHeadless.class)
+    static final class Target_java_awt_Window {
+
+        @Substitute
+        void dispose() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        void doDispose() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        void closeSplashScreen() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.awt.X11.XWindow", onlyWith = IsHeadless.class)
+    static final class Target_sun_awt_X11_XWindow {
+
+        @Substitute
+        private static void initIDs() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.awt.X11.XBaseWindow", onlyWith = IsHeadless.class)
+    @Delete
+    static final class Target_sun_awt_X11_XBaseWindow {
+    }
+
+    @TargetClass(className = "sun.awt.X11.XlibWrapper", onlyWith = IsHeadless.class)
+    static final class Target_sun_awt_X11_XlibWrapper {
+
+        @Substitute
+        static long DisplayWidth(long display, long screen) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static long DefaultScreen(long display) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static long DisplayWidthMM(long display, long screen) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static long InternAtom(long display, String string, int only_if_exists) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static void XSync(long display, int discard) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static long XAllocSizeHints() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static String[] XTextPropertyToStringList(byte[] bytes, long encoding_atom) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static String XGetAtomName(long display, long atom) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static byte[] getStringBytes(long str_ptr) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static long XAllocWMHints() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static void XFree(long ptr) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static void memcpy(long dest_ptr, long src_ptr, long length) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        static long SetToolkitErrorHandler() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.java2d.xr.XRBackendNative", onlyWith = IsHeadless.class)
+    static final class Target_sun_java2d_xr_XRBackendNative {
+
+        @Substitute
+        static void initIDs() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private static void XRAddGlyphsNative(int glyphSet,
+                        long[] glyphInfoPtrs,
+                        int glyphCnt,
+                        byte[] pixelData,
+                        int pixelDataLength) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private static int XRCreateLinearGradientPaintNative(float[] fractionsArray,
+                        short[] pixelsArray,
+                        int x1, int y1, int x2, int y2,
+                        int numStops, int repeat) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private static void XRSetClipNative(long dst,
+                        int x1, int y1, int x2, int y2,
+                        sun.java2d.pipe.Region clip, boolean isGC) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private static void XRenderCompositeTextNative(int op, int src, int dst,
+                        int srcX, int srcY, long maskFormat,
+                        int[] eltArray, int[] glyphIDs, int eltCnt,
+                        int glyphCnt) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private static void GCRectanglesNative(int drawable, long gc,
+                        int[] rectArray, int rectCnt) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private static void XRFreeGlyphsNative(int glyphSet,
+                        int[] gids, int idCnt) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private static void XRenderRectanglesNative(int dst, byte op,
+                        short red, short green,
+                        short blue, short alpha,
+                        int[] rects, int rectCnt) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private static int XRCreateRadialGradientPaintNative(float[] fractionsArray,
+                        short[] pixelsArray, int numStops,
+                        int centerX, int centerY,
+                        int innerRadius, int outerRadius,
+                        int repeat) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void renderComposite(byte op, int src, int mask,
+                        int dst, int srcX, int srcY,
+                        int maskX, int maskY, int dstX, int dstY,
+                        int width, int height) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void setGCMode(long gc, boolean copy) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void freePicture(int picture) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void setPictureRepeat(int picture, int repeat) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void setGCForeground(long gc, int pixel) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void setFilter(int picture, int filter) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private void renderRectangle(int dst, byte op,
+                        short red, short green,
+                        short blue, short alpha,
+                        int x, int y, int width, int height) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private void XRSetTransformNative(int pic,
+                        int m00, int m01, int m02,
+                        int m10, int m11, int m12) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void setGCExposures(long gc, boolean exposure) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void copyArea(int src, int dst, long gc,
+                        int srcx, int srcy, int width, int height,
+                        int dstx, int dsty) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.awt.X11InputMethodBase", onlyWith = IsHeadless.class)
+    static final class Target_sun_awt_X11InputMethodBase {
+
+        @Substitute
+        static void initIDs() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        void turnoffStatusWindow() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        boolean setCompositionEnabledNative(boolean enable) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        boolean isCompositionEnabledNative() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        String resetXIC() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        void disposeXIC() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.awt.UNIXToolkit", onlyWith = IsHeadless.class)
+    static final class Target_sun_awt_UNIXToolkit {
+
+        @Substitute
+        private static boolean load_gtk(int version, boolean verbose) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void sync() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public java.awt.image.BufferedImage getStockIcon(final int widgetType, final String stockId,
+                        final int iconSize, final int direction,
+                        final String detail) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public boolean isNativeGTKAvailable() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private boolean gtkCheckVersionImpl(int major, int minor, int micro) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.awt.X11GraphicsConfig", onlyWith = IsHeadless.class)
+    static final class Target_sun_awt_X11GraphicsConfig {
+
+        @Substitute
+        static void initIDs() {
+            throw new UnsupportedOperationException();
+        }
+
+    }
+
+    @TargetClass(className = "java.awt.AWTEvent", onlyWith = IsHeadless.class)
+    static final class Target_java_awt_AWTEvent {
+
+        @Substitute
+        private void nativeSetSource(java.awt.peer.ComponentPeer peer) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.java2d.opengl.OGLSurfaceData", onlyWith = IsHeadless.class)
+    static final class Target_sun_java2d_opengl_OGLSurfaceData {
+
+        @Substitute
+        protected boolean initFlipBackbuffer(long pData) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.java2d.opengl.OGLRenderQueue", onlyWith = IsHeadless.class)
+    static final class Target_sun_java2d_opengl_OGLRenderQueue {
+
+        @Substitute
+        private void flushBuffer(long buf, int limit) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.awt.X11.XTaskbarPeer", onlyWith = IsHeadless.class)
+    static final class Target_sun_awt_X11_XTaskbarPeer {
+
+        @Substitute
+        private static void initWithLock() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.awt.X11.XDesktopPeer", onlyWith = IsHeadless.class)
+    static final class Target_sun_awt_X11_XDesktopPeer {
+
+        @Substitute
+        private static void initWithLock() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.awt.X11.GtkFileDialogPeer", onlyWith = IsHeadless.class)
+    static final class Target_sun_awt_X11_GtkFileDialogPeer {
+
+        @Substitute
+        private static void initIDs() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private void quit() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        private void run(String title, int mode, String dir, String file,
+                        FilenameFilter filter, boolean isMultipleMode, int x, int y) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void setBounds(int x, int y, int width, int height, int op) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Substitute
+        public void toFront() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    @TargetClass(className = "sun.awt.X11.XRobotPeer", onlyWith = IsHeadless.class)
+    static final class Target_sun_awt_X11_XRobotPeer {
+
+        @Substitute
+        private static void loadNativeLibraries() {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    // To support headless mode
+    static class IsHeadless implements BooleanSupplier {
+        @Override
+        public boolean getAsBoolean() {
+            return GraphicsEnvironment.isHeadless() && JavaVersionUtil.JAVA_SPEC >= 11;
+        }
+    }
+    // Checkstyle: resume
 }

--- a/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/SystemPropertiesSupport.java
+++ b/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/SystemPropertiesSupport.java
@@ -30,6 +30,8 @@ import java.util.Map;
 import java.util.Properties;
 import java.util.function.Supplier;
 
+import com.oracle.svm.core.OS;
+import org.graalvm.compiler.serviceprovider.JavaVersionUtil;
 import org.graalvm.nativeimage.ImageInfo;
 import org.graalvm.nativeimage.Platform;
 import org.graalvm.nativeimage.Platforms;
@@ -99,6 +101,8 @@ public abstract class SystemPropertiesSupport {
         initializeProperty("java.ext.dirs", "");
         initializeProperty("java.library.path", "");
         initializeProperty("sun.arch.data.model", Integer.toString(ConfigurationValues.getTarget().wordJavaKind.getBitCount()));
+        // Some JDK libraries throws an error if 'java.home' is null.
+        initializeProperty("java.home", "undefined");
 
         String targetName = System.getProperty("svm.targetName");
         String targetArch = System.getProperty("svm.targetArch");
@@ -106,6 +110,12 @@ public abstract class SystemPropertiesSupport {
         initializeProperty("os.arch", targetArch != null ? targetArch : System.getProperty("os.arch"));
 
         initializeProperty(ImageInfo.PROPERTY_IMAGE_CODE_KEY, ImageInfo.PROPERTY_IMAGE_CODE_VALUE_RUNTIME);
+
+        if (OS.getCurrent() == OS.LINUX || JavaVersionUtil.JAVA_SPEC >= 11) {
+            initializeProperty("awt.toolkit", "sun.awt.X11.XToolkit");
+            initializeProperty("java.awt.graphicsenv", "sun.awt.X11GraphicsEnvironment");
+            initializeProperty("java.awt.printerjob", "sun.print.PSPrinterJob");
+        }
 
         lazyRuntimeValues = new HashMap<>();
         lazyRuntimeValues.put("user.name", this::userName);

--- a/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java
+++ b/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java
@@ -732,6 +732,13 @@ public class NativeImage {
         addImageBuilderJavaArgs("-Dorg.graalvm.config=" + graalvmConfig);
         addImageBuilderJavaArgs("-Dcom.oracle.graalvm.isaot=true");
         addImageBuilderJavaArgs("-Djava.system.class.loader=" + CUSTOM_SYSTEM_CLASS_LOADER);
+
+        if (OS.getCurrent() == OS.LINUX || JavaVersionUtil.JAVA_SPEC >= 11) {
+            addImageBuilderJavaArgs("-Dawt.toolkit=sun.awt.X11.XToolkit");
+            addImageBuilderJavaArgs("-Djava.awt.graphicsenv=sun.awt.X11GraphicsEnvironment");
+            addImageBuilderJavaArgs("-Djava.awt.printerjob=sun.print.PSPrinterJob");
+        }
+
         /*
          * The presence of CDS and custom system class loaders disables the use of archived
          * non-system class and and triggers a warning.

--- a/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JDKRegistrations.java
+++ b/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JDKRegistrations.java
@@ -25,10 +25,8 @@
 package com.oracle.svm.hosted.jdk;
 
 import org.graalvm.compiler.serviceprovider.JavaVersionUtil;
-import org.graalvm.nativeimage.ImageSingletons;
 
 import com.oracle.svm.core.annotate.AutomaticFeature;
-import com.oracle.svm.core.configure.ResourcesRegistry;
 import com.oracle.svm.core.graal.GraalFeature;
 import com.oracle.svm.core.jdk.JNIRegistrationUtil;
 
@@ -50,21 +48,4 @@ class JDKRegistrations extends JNIRegistrationUtil implements GraalFeature {
             rerunClassInit(a, "java.lang.ProcessImpl", "java.lang.ProcessHandleImpl", "java.lang.ProcessHandleImpl$Info", "java.io.FilePermission");
         }
     }
-
-    @Override
-    public void beforeAnalysis(BeforeAnalysisAccess access) {
-        if (JavaVersionUtil.JAVA_SPEC > 8) {
-            access.registerReachabilityHandler(this::registerColorProfileResources, clazz(access, "java.awt.color.ICC_Profile"));
-
-            /* These classes contain standard color profile caches which may not be loaded yet */
-            rerunClassInit(access, "java.awt.color.ColorSpace");
-            rerunClassInit(access, "java.awt.color.ICC_Profile");
-        }
-    }
-
-    public void registerColorProfileResources(@SuppressWarnings("unused") DuringAnalysisAccess access) {
-        ResourcesRegistry resourcesRegistry = ImageSingletons.lookup(ResourcesRegistry.class);
-        resourcesRegistry.addResources("sun.java2d.cmm.profiles.*");
-    }
-
 }

--- a/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationAwt.java
+++ b/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationAwt.java
@@ -48,7 +48,7 @@ public class JNIRegistrationAwt extends JNIRegistrationUtil implements Feature {
 
     @Override
     public void beforeAnalysis(BeforeAnalysisAccess access) {
-        if (JavaVersionUtil.JAVA_SPEC >= 11 && Platform.includedIn(Platform.LINUX_AMD64.class)) {
+        if (JavaVersionUtil.JAVA_SPEC >= 11 && Platform.includedIn(Platform.LINUX.class)) {
             access.registerReachabilityHandler(JNIRegistrationAwt::handlePreferencesClassReachable,
                             clazz(access, "java.awt.Toolkit"),
                             clazz(access, "sun.java2d.cmm.lcms.LCMS"),

--- a/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationAwt.java
+++ b/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationAwt.java
@@ -1,0 +1,207 @@
+/*
+ * Copyright (c) 2020, 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.hosted.jdk;
+
+import com.oracle.svm.core.annotate.AutomaticFeature;
+import com.oracle.svm.core.configure.ResourcesRegistry;
+import com.oracle.svm.core.jdk.JNIRegistrationUtil;
+import com.oracle.svm.core.jdk.NativeLibrarySupport;
+import com.oracle.svm.core.jdk.PlatformNativeLibrarySupport;
+import com.oracle.svm.core.jni.JNIRuntimeAccess;
+import com.oracle.svm.hosted.FeatureImpl;
+import com.oracle.svm.hosted.c.NativeLibraries;
+import org.graalvm.compiler.serviceprovider.JavaVersionUtil;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.impl.InternalPlatform;
+
+import java.awt.GraphicsEnvironment;
+
+@Platforms({InternalPlatform.PLATFORM_JNI.class})
+@AutomaticFeature
+@SuppressWarnings({"unused"})
+public class JNIRegistrationAwt extends JNIRegistrationUtil implements Feature {
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        if (JavaVersionUtil.JAVA_SPEC >= 11 && Platform.includedIn(Platform.LINUX_AMD64.class)) {
+            access.registerReachabilityHandler(JNIRegistrationAwt::handlePreferencesClassReachable,
+                            clazz(access, "java.awt.Toolkit"),
+                            clazz(access, "sun.java2d.cmm.lcms.LCMS"),
+                            clazz(access, "java.awt.event.NativeLibLoader"),
+                            clazz(access, "sun.awt.NativeLibLoader"),
+                            clazz(access, "sun.awt.image.NativeLibLoader"),
+                            clazz(access, "java.awt.image.ColorModel"),
+                            clazz(access, "sun.awt.X11GraphicsEnvironment"),
+                            clazz(access, "sun.font.FontManagerNativeLibrary"),
+                            clazz(access, "sun.java2d.Disposer"));
+
+            access.registerReachabilityHandler(JNIRegistrationAwt::registerFreeType,
+                            clazz(access, "sun.font.FontManagerNativeLibrary"));
+
+            access.registerReachabilityHandler(JNIRegistrationAwt::registerLCMS,
+                            clazz(access, "sun.java2d.cmm.lcms.LCMS"));
+
+            access.registerReachabilityHandler(JNIRegistrationAwt::registerImagingLib,
+                            clazz(access, "sun.awt.image.ImagingLib"));
+
+            access.registerReachabilityHandler(JNIRegistrationAwt::registerJPEG,
+                            clazz(access, "sun.awt.image.JPEGImageDecoder"),
+                            clazz(access, "com.sun.imageio.plugins.jpeg.JPEGImageReader"),
+                            clazz(access, "com.sun.imageio.plugins.jpeg.JPEGImageWriter"));
+
+            access.registerReachabilityHandler(JNIRegistrationAwt::registerColorProfiles,
+                            clazz(access, "java.awt.color.ICC_Profile"));
+
+            access.registerReachabilityHandler(JNIRegistrationAwt::registerFlavorMapProps,
+                            clazz(access, "java.awt.datatransfer.SystemFlavorMap"));
+
+            access.registerReachabilityHandler(JNIRegistrationAwt::registerRTFReaderCharsets,
+                            clazz(access, "javax.swing.text.rtf.RTFReader"));
+
+            access.registerReachabilityHandler(JNIRegistrationAwt::registerOceanThemeIcons,
+                            clazz(access, "javax.swing.plaf.metal.OceanTheme"));
+        }
+    }
+
+    private static void handlePreferencesClassReachable(DuringAnalysisAccess access) {
+
+        JNIRuntimeAccess.register(method(access, "java.lang.System", "setProperty", String.class, String.class));
+        JNIRuntimeAccess.register(method(access, "java.lang.System", "loadLibrary", String.class));
+
+        JNIRuntimeAccess.register(java.awt.GraphicsEnvironment.class);
+        JNIRuntimeAccess.register(method(access, "java.awt.GraphicsEnvironment", "isHeadless"));
+
+        NativeLibraries nativeLibraries = getNativeLibraries(access);
+
+        NativeLibrarySupport.singleton().preregisterUninitializedBuiltinLibrary("awt");
+        nativeLibraries.addStaticJniLibrary("awt");
+
+        if (isHeadless()) {
+            NativeLibrarySupport.singleton().preregisterUninitializedBuiltinLibrary("awt_headless");
+            nativeLibraries.addStaticJniLibrary("awt_headless", "awt");
+        } else {
+            NativeLibrarySupport.singleton().preregisterUninitializedBuiltinLibrary("awt_xawt");
+            nativeLibraries.addStaticJniLibrary("awt_xawt", "awt");
+
+            nativeLibraries.addDynamicNonJniLibrary("X11");
+            nativeLibraries.addDynamicNonJniLibrary("Xrender");
+            nativeLibraries.addDynamicNonJniLibrary("Xext");
+            nativeLibraries.addDynamicNonJniLibrary("Xi");
+        }
+
+        PlatformNativeLibrarySupport.singleton().addBuiltinPkgNativePrefix("java_awt");
+        PlatformNativeLibrarySupport.singleton().addBuiltinPkgNativePrefix("sun_awt");
+        PlatformNativeLibrarySupport.singleton().addBuiltinPkgNativePrefix("sun_java2d");
+
+        nativeLibraries.addDynamicNonJniLibrary("stdc++");
+        nativeLibraries.addDynamicNonJniLibrary("m");
+
+        access.registerReachabilityHandler(JNIRegistrationAwt::registerHtml32bdtd,
+                        clazz(access, "javax.swing.text.html.HTMLEditorKit"));
+    }
+
+    private static void registerJPEG(DuringAnalysisAccess access) {
+        NativeLibraries nativeLibraries = getNativeLibraries(access);
+
+        NativeLibrarySupport.singleton().preregisterUninitializedBuiltinLibrary("javajpeg");
+        nativeLibraries.addStaticJniLibrary("javajpeg");
+    }
+
+    private static void registerImagingLib(DuringAnalysisAccess access) {
+        NativeLibraries nativeLibraries = getNativeLibraries(access);
+
+        NativeLibrarySupport.singleton().preregisterUninitializedBuiltinLibrary("mlib_image");
+        nativeLibraries.addStaticJniLibrary("mlib_image");
+    }
+
+    private static void registerLCMS(DuringAnalysisAccess access) {
+        NativeLibraries nativeLibraries = getNativeLibraries(access);
+
+        NativeLibrarySupport.singleton().preregisterUninitializedBuiltinLibrary("lcms");
+        nativeLibraries.addStaticJniLibrary("lcms");
+    }
+
+    private static void registerFreeType(DuringAnalysisAccess access) {
+        NativeLibraries nativeLibraries = getNativeLibraries(access);
+
+        NativeLibrarySupport.singleton().preregisterUninitializedBuiltinLibrary("fontmanager");
+        nativeLibraries.addStaticJniLibrary("fontmanager", isHeadless() ? "awt_headless" : "awt_xawt");
+
+        PlatformNativeLibrarySupport.singleton().addBuiltinPkgNativePrefix("sun_font");
+
+        nativeLibraries.addDynamicNonJniLibrary("freetype");
+
+        JNIRuntimeAccess.register(clazz(access, "sun.font.FontConfigManager$FontConfigInfo"));
+        JNIRuntimeAccess.register(fields(access, "sun.font.FontConfigManager$FontConfigInfo", "fcVersion", "cacheDirs"));
+        JNIRuntimeAccess.register(clazz(access, "sun.font.FontConfigManager$FcCompFont"));
+        JNIRuntimeAccess.register(fields(access, "sun.font.FontConfigManager$FcCompFont", "fcName", "firstFont", "allFonts"));
+        JNIRuntimeAccess.register(clazz(access, "sun.font.FontConfigManager$FontConfigFont"));
+        JNIRuntimeAccess.register(constructor(access, "sun.font.FontConfigManager$FontConfigFont"));
+        JNIRuntimeAccess.register(fields(access, "sun.font.FontConfigManager$FontConfigFont", "familyName", "styleStr", "fullName", "fontFile"));
+    }
+
+    private static void registerColorProfiles(DuringAnalysisAccess duringAnalysisAccess) {
+        ResourcesRegistry resourcesRegistry = ImageSingletons.lookup(ResourcesRegistry.class);
+        resourcesRegistry.addResources("sun.java2d.cmm.profiles.*");
+    }
+
+    private static void registerFlavorMapProps(DuringAnalysisAccess duringAnalysisAccess) {
+        ResourcesRegistry resourcesRegistry = ImageSingletons.lookup(ResourcesRegistry.class);
+        resourcesRegistry.addResources("sun.datatransfer.resources.flavormap.properties");
+    }
+
+    private static void registerRTFReaderCharsets(DuringAnalysisAccess duringAnalysisAccess) {
+        ResourcesRegistry resourcesRegistry = ImageSingletons.lookup(ResourcesRegistry.class);
+        resourcesRegistry.addResources("javax.swing.text.rtf.charsets.*");
+    }
+
+    private static void registerOceanThemeIcons(DuringAnalysisAccess duringAnalysisAccess) {
+        ResourcesRegistry resourcesRegistry = ImageSingletons.lookup(ResourcesRegistry.class);
+        resourcesRegistry.addResources("javax.swing.plaf.metal.icons.*");
+        resourcesRegistry.addResources("javax.swing.plaf.basic.icons.*");
+    }
+
+    private static void registerHtml32bdtd(DuringAnalysisAccess duringAnalysisAccess) {
+        ResourcesRegistry resourcesRegistry = ImageSingletons.lookup(ResourcesRegistry.class);
+        resourcesRegistry.addResources("javax.swing.text.html.parser.html32.bdtd");
+    }
+
+    private static void registerDefaultCSS(DuringAnalysisAccess duringAnalysisAccess) {
+        ResourcesRegistry resourcesRegistry = ImageSingletons.lookup(ResourcesRegistry.class);
+        resourcesRegistry.addResources("javax.swing.text.html.default.css");
+    }
+
+    private static NativeLibraries getNativeLibraries(DuringAnalysisAccess access) {
+        FeatureImpl.DuringAnalysisAccessImpl a = (FeatureImpl.DuringAnalysisAccessImpl) access;
+        return a.getNativeLibraries();
+    }
+
+    private static boolean isHeadless() {
+        return GraphicsEnvironment.isHeadless();
+    }
+}

--- a/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/snippets/SubstrateGraphBuilderPlugins.java
+++ b/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/snippets/SubstrateGraphBuilderPlugins.java
@@ -24,6 +24,7 @@
  */
 package com.oracle.svm.hosted.snippets;
 
+import java.awt.GraphicsEnvironment;
 import java.lang.reflect.Array;
 import java.lang.reflect.Field;
 import java.lang.reflect.InvocationHandler;
@@ -159,6 +160,16 @@ class Options {
 }
 
 public class SubstrateGraphBuilderPlugins {
+    private static final String reflectionClass;
+
+    static {
+        if (JavaVersionUtil.JAVA_SPEC <= 8) {
+            reflectionClass = "sun.reflect.Reflection";
+        } else {
+            reflectionClass = "jdk.internal.reflect.Reflection";
+        }
+    }
+
     public static void registerInvocationPlugins(AnnotationSubstitutionProcessor annotationSubstitutions, MetaAccessProvider metaAccess,
                     SnippetReflectionProvider snippetReflection, InvocationPlugins plugins, Replacements replacements, boolean analysis) {
 
@@ -180,6 +191,7 @@ public class SubstrateGraphBuilderPlugins {
         registerJFREventTokenPlugins(plugins, replacements);
         registerVMConfigurationPlugins(snippetReflection, plugins);
         registerPlatformPlugins(snippetReflection, plugins);
+        registerAWTPlugins(plugins);
         registerSizeOfPlugins(snippetReflection, plugins);
         registerReferenceAccessPlugins(plugins);
     }
@@ -195,16 +207,6 @@ public class SubstrateGraphBuilderPlugins {
                     return true;
                 }
             });
-        }
-    }
-
-    private static final String reflectionClass;
-
-    static {
-        if (JavaVersionUtil.JAVA_SPEC <= 8) {
-            reflectionClass = "sun.reflect.Reflection";
-        } else {
-            reflectionClass = "jdk.internal.reflect.Reflection";
         }
     }
 
@@ -1024,6 +1026,23 @@ public class SubstrateGraphBuilderPlugins {
                 Class<? extends Platform> platform = constantObjectParameter(b, snippetReflection, targetMethod, 0, Class.class, classNode);
                 boolean result = Platform.includedIn(platform);
                 b.addPush(JavaKind.Boolean, ConstantNode.forBoolean(result));
+                return true;
+            }
+        });
+    }
+
+    /*
+     * To prevent AWT linkage error that happens with 'awt_headless' in headless mode, we eliminate
+     * native methods that depend on 'awt_xawt' library in the call-tree.
+     */
+    private static void registerAWTPlugins(InvocationPlugins plugins) {
+        Registration r = new Registration(plugins, GraphicsEnvironment.class);
+        boolean isHeadless = GraphicsEnvironment.isHeadless();
+        r.register0("isHeadless", new InvocationPlugin() {
+            @SuppressWarnings("unchecked")
+            @Override
+            public boolean apply(GraphBuilderContext b, ResolvedJavaMethod targetMethod, Receiver receiver) {
+                b.addPush(JavaKind.Boolean, ConstantNode.forBoolean(isHeadless));
                 return true;
             }
         });

--- a/substratevm/src/com.oracle.svm.jni/src/com/oracle/svm/jni/JNILibraryInitializer.java
+++ b/substratevm/src/com.oracle.svm.jni/src/com/oracle/svm/jni/JNILibraryInitializer.java
@@ -82,7 +82,9 @@ public class JNILibraryInitializer implements NativeLibrarySupport.LibraryInitia
     }
 
     public boolean fillCGlobalDataMap(Collection<String> staticLibNames) {
-        List<String> libsWithOnLoad = Arrays.asList("net", "java", "nio", "zip", "sunec", "jaas", "sctp", "extnet", "j2gss", "j2pkcs11", "j2pcsc", "prefs", "verify");
+        List<String> libsWithOnLoad = Arrays.asList("net", "java", "nio", "zip", "sunec", "jaas", "sctp", "extnet",
+                        "j2gss", "j2pkcs11", "j2pcsc", "prefs", "verify", "awt", "awt_xawt", "awt_headless", "lcms",
+                        "fontmanager", "javajpeg", "mlib_image");
         // TODO: This check should be removed when all static libs will have JNI_OnLoad function
         ArrayList<String> localStaticLibNames = new ArrayList<>(staticLibNames);
         localStaticLibNames.retainAll(libsWithOnLoad);

